http-response/background.js

```diff
@@ -1,22 +1,154 @@
 }
 
-browser.webRequest.onBeforeRequest.addListener(
+browser.webRequest.onHeadersReceived.addListener(
   listener,
-  {urls: ["https://example.com/*"], types: ["main_frame"]},
-  ["blocking"]
+  {
+    urls: ["https://example.com/*", "https://www.w3.org/*"], // Include W3 for testing charset detection.
+    types: ["main_frame"]
+  },
+  ["blocking","responseHeaders"]
 );
+
+// This helper method does a few things regarding character encoding:
+// 1) Detects the charset for the TextDecoder so that bytes are properly turned into strings
+// 2) Ensures the output Content-Type is UTF-8 because that is what TextEncoder supports
+// 3) Returns the decoder/encoder pair
+function detectCharsetAndSetupDecoderEncoder(details) {
+  let contentType = '';
+  let headerIndex = -1;
+  for(let i=0; i<details.responseHeaders.length; i++) {
+      let header = details.responseHeaders[i];
+      if(header.name.toLowerCase() == "content-type") {
+          contentType = header.value.toLowerCase();
+          headerIndex = i;
+          break;
+      }
+  }
+  if (headerIndex == -1) {
+    console.log('No Content-Type header detected for '+details.url+', adding one.');
+    headerIndex = details.responseHeaders.length;
+    contentType = 'text/html';
+    details.responseHeaders.push(
+      {
+        "name": "Content-Type",
+        "value":"text/html"
+      }
+    );
+  }
+
+  let baseType;
+  if(contentType.trim().startsWith('text/html')) {
+    baseType = 'text/html';
+    console.log('Detected base type was '+baseType);
+  } else if(contentType.trim().startsWith('application/xhtml+xml')) {
+    baseType = 'application/xhtml+xml';
+    console.log('Detected base type was '+baseType);
+  } else {
+    baseType = 'text/html';
+    console.log('The Content-Type was '+contentType+', not text/html or application/xhtml+xml - results might be strange.');
+  }
+
+  // It is important to detect the charset to correctly initialize TextDecoder or
+  // else we run into garbage output sometimes.
+  // However, TextEncoder does NOT support other than 'utf-8', so it is necessary
+  // to change the Content-Type on the header to UTF-8
+  // If modifying this block of code, ensure that the tests at
+  // https://www.w3.org/2006/11/mwbp-tests/index.xhtml
+  // all pass - current implementation only fails on #9 but this detection ensures
+  // tests #3,4,5, and 8 pass.
+  let decodingCharset = 'utf-8';
+  let detectedCharset = detectCharset(contentType);
+
+  if(detectedCharset !== undefined) {
+      decodingCharset = detectedCharset;
+      console.log('Detected charset was ' + decodingCharset + ' for ' + details.url);
+  }
+  details.responseHeaders[headerIndex].value = baseType+';charset=utf-8';
+
+  let decoder = new TextDecoder(decodingCharset);
+  let encoder = new TextEncoder(); //Encoder does not support non-UTF-8 charsets so this is always utf-8.
+
+  return [decoder,encoder];
+}
+
+// Detect the charset from Content-Type
+function detectCharset(contentType) {

```

Comment:
The charset detection logic as implemented below is incorrect (e.g. when multiple `charset` parameters are set, the last one should be used; when multiple MIME types are set, the previously set charset should be forgotten). Let's keep the example simple, however, and make it more obvious that the implementation is incomplete.

Here is an example of a more complete implementation that also covers some edge cases: https://github.com/Rob--W/open-in-browser/commit/a6b926ea9522b35298632e5e6a2c89ddb456c5d9

As you can see, it's quite involved - do not put the code in the PR. You can link to it if you wish though.

Reply:
There's wisdom here; thanks for the link.
I've been thinking about your comment about an obviously incomplete implementation. While the idea of a complete solution here is appealing, the simplified solution will likely cover 99% of the cases and the comment will make the careful reader aware of the problem. Lack of awareness of this problem - albeit obvious in retrospect - is why I'm wanting to contribute back a more robust example in the first place; a user reported a bug to me in a review of my addon. I'd like others to avoid the same fate, as their users may not be as considerate as mine was.

- [ ] Make incompleteness obvious

Given both the complexity of determining the charset as well as the need for it to be accurate in order to filter text properly, do you think the browser-predicted charset would ever be a candidate to expose on the `details` object? It just seems like right now it would be incredibly difficult to create a truly correct implementation - do you know of any extant addons that are actually correct in this regard?

Reply:
> Given both the complexity of determining the charset as well as the need for it to be accurate in order to filter text properly, do you think the browser-predicted charset would ever be a candidate to expose on the `details` object?

What a good idea. I have thought of it before and posted a feature request at https://bugzilla.mozilla.org/show_bug.cgi?id=1425479
Unfortunately the feature request was denied. If you post a new bug that shows use cases and why alternatives aren't a good fit, the feature request can be re-evaluated.

>  It just seems like right now it would be incredibly difficult to create a truly correct implementation - do you know of any extant addons that are actually correct in this regard?

The Open in Browser add-on that I linked before was authored by me, and modelled after Firefox's implementation (which includes non-standard behavior).